### PR TITLE
Allow cascade deletion of auditlog entries

### LINKED ISSUE #493: Can't remove any auditlogged objects through the admin
As @madomdy [commented](https://github.com/jazzband/django-auditlog/pull/449#issuecomment-1370397139) in a related pull request, they are unable to remove objects of any model connected to the audit log.

screenshot

Possibly minimal app to reproduce the issue (I haven't tested it):
```python
# model.py
from auditlog.models import AuditlogHistoryField
from django.db import models


class MyModel(models.Model):
    history = AuditlogHistoryField()    

# admin.py
from django.contrib import admin

from .models import MyModel


admin.site.register(MyModel)
```

Instances of `MyModel` can't be removed through the admin because of `AuditlogHistoryField` and the delete permissions retracted in #449.

Workaround: pass `delete_related=False` to `AuditlogHistoryField` (`history = AuditlogHistoryField(delete_related=False)`).

### PR DESCRIPTION
**Pull request by aleh-rymasheuski**
Resolves #493.

This pull request allows cascade deletion of log entries, but still disallows direct deletion through the admin.

### CODE REVIEW

**Comment by aleh-rymasheuski** [comment]
@hramezani, we don't have a structure for testing this case with automated tests, and I don't have an elegant way to test this. If you can think of a good way of testing this, please add a test. Otherwise I'd suggest merging this change without a test.

File: auditlog/admin.py [hunk 1]
```diff
@@ -40,6 +40,11 @@ def has_change_permission(self, request, obj=None):
         return False
 
     def has_delete_permission(self, request, obj=None):
+        if request.resolver_match.url_name not in [
+            pattern.name for pattern in self.urls if pattern.name
+        ]:
```
Comment (hramezani) [inline auditlog/admin.py:45]: @aleh-rymasheuski Could you please explain more about what is happening here?

Also, I think you can move the `[pattern.name for pattern in self.urls if pattern.name]` out of method and define a class attr to don't repeat the loop in each call.

**Comment by hramezani** [comment]
> @hramezani, we don't have a structure for testing this case with automated tests, and I don't have an elegant way to test this. If you can think of a good way of testing this, please add a test. Otherwise I'd suggest merging this change without a test.

Thanks @aleh-rymasheuski for this patch 🙏 

I think we can test it like:

```diff
diff --git a/auditlog_tests/tests.py b/auditlog_tests/tests.py
index 12a4725..8824780 100644
--- a/auditlog_tests/tests.py
+++ b/auditlog_tests/tests.py
@@ -19,7 +19,7 @@ from django.core import management
 from django.db import models
 from django.db.models.signals import pre_save
 from django.test import RequestFactory, TestCase, override_settings
-from django.urls import reverse
+from django.urls import resolve, reverse
 from django.utils import dateformat, formats
 from django.utils import timezone as django_timezone
 from django.utils.encoding import smart_str
@@ -1555,6 +1555,9 @@ class DiffMsgTest(TestCase):
         super().setUp()
         self.site = AdminSite()
         self.admin = LogEntryAdmin(LogEntry, self.site)
+        self.user = User.objects.create_user(
+            username="test", email="test@example.com", password="top_secret"
+        )
 
     def _create_log_entry(self, action, changes):
         return LogEntry.objects.log_create(
@@ -1711,6 +1714,13 @@ class DiffMsgTest(TestCase):
         ):
             self.assertEqual(self.admin.field_verbose_name(log_entry, "actor"), "actor")
 
+    def test_has_delete_permission(self):
+        request = RequestFactory().get("/simplemodel/1/")
+        request.resolver_match = resolve("/simplemodel/1/")
+        request.user = self.user
+
+        self.assertFalse(self.admin.has_delete_permission(request))
+
 
 class NoDeleteHistoryTest(TestCase):
     def test_delete_related(self):
```

**Comment by aleh-rymasheuski** [comment]
@hramezani, I added a test based on your suggestion.

File: auditlog/admin.py [hunk 2]
```diff
@@ -40,6 +40,11 @@ def has_change_permission(self, request, obj=None):
         return False
 
     def has_delete_permission(self, request, obj=None):
+        if request.resolver_match and request.resolver_match.url_name not in [
+            pattern.name for pattern in self.urls if pattern.name
```
Comment (hramezani) [inline auditlog/admin.py:44]: can't we move the admin urls patterns list creation out of this method to not calculate it every time?

Comment (aleh-rymasheuski) [reply auditlog/admin.py:44]: I'm not sure when in the lifecycle of the admin `self.urls` is available, so I'll just move this expression into a cached property.

Comment (aleh-rymasheuski) [reply auditlog/admin.py]: Done.

**Review by hramezani — APPROVED** [review]

**Pull request merged**